javascript/node/selenium-webdriver/net/portprober.js

```diff
@@ -161,15 +161,23 @@ function isFree(port, opt_host) {
  * @return {!Promise<number>} A promise that will resolve to a free port. If a
  *     port cannot be found, the promise will be rejected.
  */
-async function findFreePort(opt_host) {
-  let range = await findSystemPortRange()
-  for (let i = 0; i < 100; i++) {
-    let port = Math.floor(Math.random() * (range.max - range.min) + range.min)
-    if (await isFree(port, opt_host)) {
-      return port
-    }
-  }
-  throw Error('Unable to find a free port')
+
+function findFreePort(opt_host) {
+  return new Promise((resolve, reject) => {
+    const server = net.createServer()
+    server.on('listening', function() {
+      resolve(server.address().port);
+      server.close();
+    });
+    server.on('error', (e) => {
+      if (e.code === 'EADDRINUSE' || e.code === 'EACCES') {
+        resolve('Unable to find a free port')
+      } else {
+        reject(e)
+      }
+    });
+    server.listen(0, opt_host);
```

Comment:
I am not a JS expert, but a few questions and comments:

- Is this guaranteeing that an ephemeral port is returned for all OS?
- This change is not returning the port anymore, was this change tested?
- What happens if the method fails? The current code retries in a range of ports.
- Would this be thread safe? If 100 tests run in parallel, will a 100 free ports be returned?
- Probably this change needs some extra tests in the `portprober_test.js`.

Reply:
1. Node documentation states the following: 
`If port is omitted or is 0, the operating system will assign an arbitrary unused port, which can be retrieved by using server.address().port after the 'listening' event has been emitted.`
2. The change is returning a port at line 169
3. It will throw an error just like it did in the old code
4. Unless there's a bug in Node or in the OS itself it should. See first point.
5. Since it still returns a port if successful or the same error message if not tests should be fine.

Reply:
Stackoverflow about how Node does this: https://stackoverflow.com/questions/9901043/how-does-node-js-choose-random-ports

Reply:
>The new code is returning a port at line 169

I completely oversaw that, thanks for pointing it out.

Then I think this makes sense, please let us know when the PR is ready for review and merge (is a draft right now).

Reply:
No worries! I guess it's ready from my sides. There was some weird issue with the test though. Some timeout in Travis.

Also ,I'm not sure how to actually test that it respects the system port range on all systems and similar stuff. 
You would have to change the system range on the system you test on and verify that it doesn't return something not in the range. 
We tried on our unix env to set net.ipv4.ip_local_port_range="1025 1026" and it always returned 1025. I could test manually on Windows as well as my Mac. But i guess it's hard to write automated test for it. 
I can also try to set the range to just 1025-1025, start a server and then try to get another port to see what actually happens when you run out.